sparse/impl/KokkosSparse_spiluk_numeric_impl.hpp

```diff
@@ -295,49 +336,111 @@ struct IlukWrap {
       KokkosBatched::TeamSetIdentity<member_type>::invoke(team, lget(block));
     }
 
-    // divide. lhs /= rhs
+    // assign
+    template <typename ViewT>
+    KOKKOS_INLINE_FUNCTION void assign(const ViewT &lhs,
+                                       const CBlock &rhs) const {
+      for (size_type i = 0; i < block_size; ++i) {
+        for (size_type j = 0; j < block_size; ++j) {
+          lhs(i, j) = rhs(i, j);
+        }
+      }
+    }
+
+    // divide. lhs /= rhs (lhs = lhs * rhs^-1)
     KOKKOS_INLINE_FUNCTION
-    void divide(const member_type &team, const LBlock &lhs,
-                const UBlock &rhs) const {
+    void divide(const member_type &team, const Block &lhs, const CBlock &rhs,
+                scalar_t *buff) const {
+      // Need a temp block to do LU of rhs
+      Block LU(buff, block_size, block_size);
+      assign(LU, rhs);
+      KokkosBatched::TeamLU<member_type,
+                            KokkosBatched::Algo::LU::Blocked>::invoke(team, LU);
+
+      // rhs = LU
+      // rhs^-1 = U^-1 * L^-1
+      // lhs = (lhs * U^-1) * L^-1, so do U trsm first
       KokkosBatched::TeamTrsm<
           member_type, KokkosBatched::Side::Right, KokkosBatched::Uplo::Upper,
-          KokkosBatched::Trans::NoTranspose,  // not 100% on this
-          KokkosBatched::Diag::NonUnit,
-          KokkosBatched::Algo::Trsm::Unblocked>::  // not 100% on this
-          invoke(team, 1.0, rhs, lhs);
+          KokkosBatched::Trans::NoTranspose, KokkosBatched::Diag::NonUnit,
+          KokkosBatched::Algo::Trsm::Blocked>::invoke(team, 1.0, LU, lhs);
+
+      KokkosBatched::TeamTrsm<
+          member_type, KokkosBatched::Side::Right, KokkosBatched::Uplo::Lower,
+          KokkosBatched::Trans::NoTranspose, KokkosBatched::Diag::Unit,
+          KokkosBatched::Algo::Trsm::Blocked>::invoke(team, 1.0, LU, lhs);
+    }
+
+    // divide_left. lhs /= rhs (lhs = rhs^-1 * lhs)
+    KOKKOS_INLINE_FUNCTION
+    void divide_left(const Block &lhs, const CBlock &rhs,
+                     scalar_t *buff) const {
+      Block LU(buff, block_size, block_size);
+      assign(LU, rhs);
+      KokkosBatched::SerialLU<KokkosBatched::Algo::LU::Blocked>::invoke(LU);
```

Comment:
Why is `divide_left` using SerialLU (and Trsm later), while `divide` uses the Team version?

Reply:
Okay, I had not seen all the context, that sounds reasonable then. Note that the MacOS builds seems to fail in the unit-test with a tolerance probably a bit too tight at least in the `float` test...